benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java

```diff
@@ -0,0 +1,237 @@
+    GrpcClientSettings settings = Utils.createGrpcClientSettings(new InetSocketAddress("127.0.0.1", port));
+    workerServiceStub = WorkerServiceClient.create(settings, mat, system.dispatcher());
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    if (system != null) {
+      workerServiceStub.close();
+      system.terminate();
+      system.getWhenTerminated().toCompletableFuture().get(10, TimeUnit.SECONDS);
+    }
+  }
+
+  @Ignore // SYNC_CLIENT not implemented
+  public void runUnaryBlockingClosedLoop() throws Exception {
+    Control.ServerArgs.Builder serverArgsBuilder = Control.ServerArgs.newBuilder();
+    serverArgsBuilder.getSetupBuilder()
+        .setServerType(Control.ServerType.ASYNC_SERVER)
+        .setAsyncServerThreads(4)
+        .setPort(0)
+        .getPayloadConfigBuilder().getSimpleParamsBuilder().setRespSize(1000);
+    int serverPort = startServer(serverArgsBuilder.build());
+
+    Control.ClientArgs.Builder clientArgsBuilder = Control.ClientArgs.newBuilder();
+    String serverAddress = "localhost:" + serverPort;
+    clientArgsBuilder.getSetupBuilder()
+        .setClientType(Control.ClientType.SYNC_CLIENT)
+        .setRpcType(Control.RpcType.UNARY)
+        .setClientChannels(2)
+        .setOutstandingRpcsPerChannel(2)
+        .addServerTargets(serverAddress);
+    clientArgsBuilder.getSetupBuilder().getPayloadConfigBuilder().getSimpleParamsBuilder()
+        .setReqSize(1000)
+        .setRespSize(1000);
+    clientArgsBuilder.getSetupBuilder().getHistogramParamsBuilder()
+        .setResolution(0.01)
+        .setMaxPossible(60000000000.0);
+    assertWorkOccurred(clientArgsBuilder.build());
+  }
+
+  @Test
+  public void runUnaryAsyncClosedLoop() throws Exception {
+    Control.ServerArgs.Builder serverArgsBuilder = Control.ServerArgs.newBuilder();
+    serverArgsBuilder.getSetupBuilder()
+        .setServerType(Control.ServerType.ASYNC_SERVER)
+        .setAsyncServerThreads(4)
+        .setPort(0)
+        .getPayloadConfigBuilder().getSimpleParamsBuilder().setRespSize(1000);
+    int serverPort = startServer(serverArgsBuilder.build());
+
+    Control.ClientArgs.Builder clientArgsBuilder = Control.ClientArgs.newBuilder();
+    String serverAddress = "localhost:" + serverPort;
+    clientArgsBuilder.getSetupBuilder()
+        .setClientType(Control.ClientType.ASYNC_CLIENT)
+        .setClientChannels(2)
+        .setRpcType(Control.RpcType.UNARY)
+        .setOutstandingRpcsPerChannel(1)
+        .setAsyncClientThreads(4)
+        .addServerTargets(serverAddress);
+    clientArgsBuilder.getSetupBuilder().getPayloadConfigBuilder().getSimpleParamsBuilder()
+        .setReqSize(1000)
+        .setRespSize(1000);
+    clientArgsBuilder.getSetupBuilder().getHistogramParamsBuilder()
+        .setResolution(0.01)
+        .setMaxPossible(60000000000.0);
+    assertWorkOccurred(clientArgsBuilder.build());
+  }
+
+  @Test
+  public void runPingPongAsyncClosedLoop() throws Exception {
```

Comment:
This test sometimes (not always) fails with:
```
[info] - runPingPongAsyncClosedLoop *** FAILED ***
[info]   java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: PROTOCOL_ERROR
[info] Received Goaway
[info] Received unexpected frame of type RstStreamFrame for stream 3 in state Closed
[info]   at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
[info]   at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.super$get(FutureConvertersImpl.scala:84)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.$anonfun$get$2(FutureConvertersImpl.scala:84)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
[info]   at scala.concurrent.package$.blocking(package.scala:142)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.get(FutureConvertersImpl.scala:84)
[info]   at akka.grpc.benchmarks.driver.LoadWorkerTest.assertWorkOccurred(LoadWorkerTest.java:221)
[info]   at akka.grpc.benchmarks.driver.LoadWorkerTest.runPingPongAsyncClosedLoop(LoadWorkerTest.java:173)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   ...
[info]   Cause: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: PROTOCOL_ERROR
```

Could that be a real issue in our implementation?

Reply:
Yes, seems like a bug: from https://http2.github.io/http2-spec/#StreamStates: "Endpoints MUST ignore WINDOW_UPDATE or RST_STREAM frames received in this state, though endpoints MAY choose to treat frames that arrive a significant time after sending END_STREAM as a connection error". Added https://github.com/akka/akka-http/issues/2087

Reply:
thanks for digging

Reply:
if it becomes to annoying we have to mark this test with `@Ignore` until fixed